### ISSUE TITLE
DataType isn't baseType for ProfileInfo

### ISSUE THREAD

**Issue by pjbss** [cqframework/clinical_quality_language#264]
Hello,

I have constructed a simple CQL statement to try and look into how NegativeEncounterPerformed would be used. Using the QDM 5.3 model the ELM is outputing the dataType as NegativeEncounterPerformed and not the defined base type of EncounterPerformed. This seems to be different that what is described in the profile type javadoc in the project, I pasted that below along with the CQL and ELM.

JavaDoc:

> Created by Bryn on 8/22/2016. Profiles within CQL allow the same underlying type to be accessed using different labels. When a profile is referenced, the ELM retrieve that is output will use the baseType of the profile as the dataType, and the name of the profile as the identifier. This allows the implementation to detect when the same data type has been accessed in a different way by the author. This mechanism is used in QDM to enable the negation pattern. For example, the Encounter, Performed data type is the same whether it is accessed positively or negatively, but the difference needs to be communicated reliably through the data access layer, so profiles are used, one positive and one negative. The underlying type Encounter, Performed, is not retrievable, only the positive and negative profiles. The identifiers are set to Encounter, Performed, and Encounter, Not Performed, and the resulting retrieve will reflect the EncounterPerformed type, together with the name of the profile, PositiveEncounterPerformed or NegativeEncounterPerformed, depending on which profile was used in the retrieve.

CQL

```
library

 AgeAtMP version '1'

using QDM

valueset "Office Visit": 'urn:oid:2.16.840.1.113883.3.464.1003.101.12.1001'

parameter MeasurementPeriod default Interval[DateTime(2013, 1, 1, 0, 0, 0, 0), DateTime(2014, 1, 1, 0, 0, 0, 0))

context Patient

define InDemographic:
    AgeInYearsAt(start of MeasurementPeriod) >= 2 and AgeInYearsAt(start of MeasurementPeriod) < 18

define "Simple Encounter":
  ["Encounter, Performed": "Office Visit"]

define "Simple Non Encounter":
  ["Encounter, Not Performed": "Office Visit"]
```

Full ELM JSON:
```
{
   "library" : {
      "identifier" : {
         "id" : "AgeAtMP",
         "version" : "1"
      },
      "schemaIdentifier" : {
         "id" : "urn:hl7-org:elm",
         "version" : "r1"
      },
      "usings" : {
         "def" : [ {
            "localIdentifier" : "System",
            "uri" : "urn:hl7-org:elm-types:r1"
         }, {
            "localId" : "1",
            "localIdentifier" : "QDM",
            "uri" : "urn:healthit-gov:qdm:v5_3"
         } ]
      },
      "parameters" : {
         "def" : [ {
            "localId" : "20",
            "name" : "MeasurementPeriod",
            "accessLevel" : "Public",
            "default" : {
               "localId" : "19",
               "lowClosed" : true,
               "highClosed" : false,
               "type" : "Interval",
               "low" : {
                  "localId" : "10",
                  "type" : "DateTime",
                  "year" : {
                     "localId" : "3",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "2013",
                     "type" : "Literal"
                  },
                  "month" : {
                     "localId" : "4",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "1",
                     "type" : "Literal"
                  },
                  "day" : {
                     "localId" : "5",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "1",
                     "type" : "Literal"
                  },
                  "hour" : {
                     "localId" : "6",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "0",
                     "type" : "Literal"
                  },
                  "minute" : {
                     "localId" : "7",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "0",
                     "type" : "Literal"
                  },
                  "second" : {
                     "localId" : "8",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "0",
                     "type" : "Literal"
                  },
                  "millisecond" : {
                     "localId" : "9",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "0",
                     "type" : "Literal"
                  }
               },
               "high" : {
                  "localId" : "18",
                  "type" : "DateTime",
                  "year" : {
                     "localId" : "11",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "2014",
                     "type" : "Literal"
                  },
                  "month" : {
                     "localId" : "12",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "1",
                     "type" : "Literal"
                  },
                  "day" : {
                     "localId" : "13",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "1",
                     "type" : "Literal"
                  },
                  "hour" : {
                     "localId" : "14",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "0",
                     "type" : "Literal"
                  },
                  "minute" : {
                     "localId" : "15",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "0",
                     "type" : "Literal"
                  },
                  "second" : {
                     "localId" : "16",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "0",
                     "type" : "Literal"
                  },
                  "millisecond" : {
                     "localId" : "17",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "0",
                     "type" : "Literal"
                  }
               }
            }
         } ]
      },
      "valueSets" : {
         "def" : [ {
            "localId" : "2",
            "name" : "Office Visit",
            "id" : "urn:oid:2.16.840.1.113883.3.464.1003.101.12.1001",
            "accessLevel" : "Public"
         } ]
      },
      "statements" : {
         "def" : [ {
            "name" : "Patient",
            "context" : "Patient",
            "expression" : {
               "type" : "SingletonFrom",
               "operand" : {
                  "dataType" : "{urn:healthit-gov:qdm:v5_3}Patient",
                  "templateId" : "Patient",
                  "type" : "Retrieve"
               }
            }
         }, {
            "localId" : "32",
            "name" : "InDemographic",
            "context" : "Patient",
            "accessLevel" : "Public",
            "annotation" : [ {
               "type" : "Annotation",
               "s" : {
                  "r" : "32",
                  "s" : [ {
                     "value" : [ "define ","InDemographic",":\n    " ]
                  }, {
                     "r" : "31",
                     "s" : [ {
                        "r" : "25",
                        "s" : [ {
                           "r" : "23",
                           "s" : [ {
                              "value" : [ "AgeInYearsAt","(" ]
                           }, {
                              "r" : "22",
                              "s" : [ {
                                 "value" : [ "start of " ]
                              }, {
                                 "r" : "21",
                                 "s" : [ {
                                    "value" : [ "MeasurementPeriod" ]
                                 } ]
                              } ]
                           }, {
                              "value" : [ ")" ]
                           } ]
                        }, {
                           "value" : [ " ",">="," ","2" ]
                        } ]
                     }, {
                        "value" : [ " and " ]
                     }, {
                        "r" : "30",
                        "s" : [ {
                           "r" : "28",
                           "s" : [ {
                              "value" : [ "AgeInYearsAt","(" ]
                           }, {
                              "r" : "27",
                              "s" : [ {
                                 "value" : [ "start of " ]
                              }, {
                                 "r" : "26",
                                 "s" : [ {
                                    "value" : [ "MeasurementPeriod" ]
                                 } ]
                              } ]
                           }, {
                              "value" : [ ")" ]
                           } ]
                        }, {
                           "value" : [ " ","<"," ","18" ]
                        } ]
                     } ]
                  } ]
               }
            } ],
            "expression" : {
               "localId" : "31",
               "type" : "And",
               "operand" : [ {
                  "localId" : "25",
                  "type" : "GreaterOrEqual",
                  "operand" : [ {
                     "localId" : "23",
                     "precision" : "Year",
                     "type" : "CalculateAgeAt",
                     "operand" : [ {
                        "path" : "birthDatetime",
                        "type" : "Property",
                        "source" : {
                           "name" : "Patient",
                           "type" : "ExpressionRef"
                        }
                     }, {
                        "localId" : "22",
                        "type" : "Start",
                        "operand" : {
                           "localId" : "21",
                           "name" : "MeasurementPeriod",
                           "type" : "ParameterRef"
                        }
                     } ]
                  }, {
                     "localId" : "24",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "2",
                     "type" : "Literal"
                  } ]
               }, {
                  "localId" : "30",
                  "type" : "Less",
                  "operand" : [ {
                     "localId" : "28",
                     "precision" : "Year",
                     "type" : "CalculateAgeAt",
                     "operand" : [ {
                        "path" : "birthDatetime",
                        "type" : "Property",
                        "source" : {
                           "name" : "Patient",
                           "type" : "ExpressionRef"
                        }
                     }, {
                        "localId" : "27",
                        "type" : "Start",
                        "operand" : {
                           "localId" : "26",
                           "name" : "MeasurementPeriod",
                           "type" : "ParameterRef"
                        }
                     } ]
                  }, {
                     "localId" : "29",
                     "valueType" : "{urn:hl7-org:elm-types:r1}Integer",
                     "value" : "18",
                     "type" : "Literal"
                  } ]
               } ]
            }
         }, {
            "localId" : "34",
            "name" : "Simple Encounter",
            "context" : "Patient",
            "accessLevel" : "Public",
            "annotation" : [ {
               "type" : "Annotation",
               "s" : {
                  "r" : "34",
                  "s" : [ {
                     "value" : [ "define ","\"Simple Encounter\"",":\n  " ]
                  }, {
                     "r" : "33",
                     "s" : [ {
                        "value" : [ "[","\"Encounter, Performed\"",": " ]
                     }, {
                        "s" : [ {
                           "value" : [ "\"Office Visit\"" ]
                        } ]
                     }, {
                        "value" : [ "]" ]
                     } ]
                  } ]
               }
            } ],
            "expression" : {
               "localId" : "33",
               "dataType" : "{urn:healthit-gov:qdm:v5_3}PositiveEncounterPerformed",
               "templateId" : "PositiveEncounterPerformed",
               "codeProperty" : "code",
               "type" : "Retrieve",
               "codes" : {
                  "name" : "Office Visit",
                  "type" : "ValueSetRef"
               }
            }
         }, {
            "localId" : "36",
            "name" : "Simple Non Encounter",
            "context" : "Patient",
            "accessLevel" : "Public",
            "annotation" : [ {
               "type" : "Annotation",
               "s" : {
                  "r" : "36",
                  "s" : [ {
                     "value" : [ "define ","\"Simple Non Encounter\"",":\n  " ]
                  }, {
                     "r" : "35",
                     "s" : [ {
                        "value" : [ "[","\"Encounter, Not Performed\"",": " ]
                     }, {
                        "s" : [ {
                           "value" : [ "\"Office Visit\"" ]
                        } ]
                     }, {
                        "value" : [ "]" ]
                     } ]
                  } ]
               }
            } ],
            "expression" : {
               "localId" : "35",
               "dataType" : "{urn:healthit-gov:qdm:v5_3}NegativeEncounterPerformed",
               "templateId" : "NegativeEncounterPerformed",
               "codeProperty" : "code",
               "type" : "Retrieve",
               "codes" : {
                  "name" : "Office Visit",
                  "type" : "ValueSetRef"
               }
            }
         } ]
      }
   }
}
```

**Comment by brynrhodes** [comment]
Hi @pjbss, this is a case of the documentation being out of date. Apologies for the confusion. This is how the retrieves and profiles were originally designed for use with QDM negation. However, when trying to use profile types with functions, we found that we needed to preserve the profile type, and having the retrieve use the base type, rather than the profile type, resulted in inconsistencies. We resolved the issue by changing the retrieve to return the profile type, rather than the base type (see issue #131 for a more detailed discussion).

I'll update the documentation on profile type to avoid this confusion in the future.

**Comment by pjbss** [comment]
Hello @brynrhodes,

Thank you for pointing me to the information about #131, that makes sense. The one remaining question I have is that I'm unable to find reference to the Negation concept or "Encounter, Not Performed" in the QDM specification PDF (QDM_5_3_ANNOTATED_2017August_508.pdf).

Is this a QDM specification, or a feature of CQL/ELM on top of the base QDM specification? I want to make sure that I fully understand what is the "QDM model".

Thank you,
Patrick

**Comment by brynrhodes** [comment]
Hi @pjbss,

Negation as a concept within QDM is modeled with the `negationRationale` attribute, and the inclusion of a negation modifier in the data type name. If you look at the documentation for this attribute in the QDM documentation, it defers the concrete syntax to an implementation guide:

> The syntax in the human readable HQMF is address in CQL examples and in the MAT User Guide. Prior versions of QDM used the syntax, “Procedure, Performed not done.” QDM 5.0 DRAFT uses the syntax, “Procedure, not Performed.”

[HL7 Version 3 Implementation Guide: Clinical Quality Language (CQL)-based Health Quality Measure Format (HQMF), Release 1 - US Realm](http://www.hl7.org/implement/standards/product_brief.cfm?product_id=405) defines the concrete syntax for negation for QDM in Volume 2, section 2.8.1, Conformance Requirement 12.

The [model info](https://github.com/cqframework/clinical_quality_language/blob/master/Src/java/qdm/src/main/resources/gov/healthit/qdm/qdm-modelinfo-5.3.xml) file for QDM version 5.3 uses these names as the identifiers for the profile types for the negation data types.

Regards,
Bryn

**Comment by brynrhodes** [comment]
Included in release 1.2.18
